lib/charms/opensearch/v0/opensearch_performance_profile.py

```diff
@@ -0,0 +1,240 @@
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 1
+
+
+logger = logging.getLogger(__name__)
+
+
+class _ApplyProfileTemplatesOpenSearch(EventBase):
+    """Attempt to acquire lock & restart OpenSearch.
+
+    The main reason to have a separate event, is to be able to wait for the cluster.
+    It defers otherwise and only defers the execution of this particular task.
+    """
+
+
+class OpenSearchPerformance(ops.Object):
+    """Base class for OpenSearch charms."""
+
+    _apply_profile_templates_event = EventSource(_ApplyProfileTemplatesOpenSearch)
+
+    def __init__(self, charm: ops.charm.CharmBase = None):
+        super().__init__(charm, None)
+        self.charm = charm
+        self.peers_data = self.charm.peers_data
+
+        self.framework.observe(
+            charm.on.install,
+            self._on_install,
+        )
+        self.framework.observe(
+            charm.on[PeerClusterRelationName].relation_changed,
+            self._on_peer_cluster_relation_changed,
+        )
+        self.framework.observe(
+            charm.on[PeerRelationName].relation_changed,
+            self._on_peer_cluster_relation_changed,
+        )
+        self.framework.observe(
+            self._apply_profile_templates_event, self._on_apply_profile_templates
+        )
+        self._apply_profile_templates_has_been_called = False
+
+    def _on_install(self, event: EventBase):
+        """Handle install event.
+
+        Execute it on install, as defined in the docs, it is the first hook to run:
+           https://juju.is/docs/sdk/config-changed-event
+
+        install -> config-changed -> start
+
+        Store the current perf. profile we are applying. This must happen solely
+        on the beginning of the charm lifecycle. Therefore, we check if the unit is down.
+        """
+        if not os.path.exists(
+            os.path.join(
+                self.charm.opensearch.paths.conf,
+                self.charm.opensearch_config.JVM_OPTIONS,
+            )
+        ):
+            event.defer()
+            return
+        self.apply(self.charm.config.get(PERFORMANCE_PROFILE))
+
+    def _on_peer_cluster_relation_changed(self, _: EventBase):
+        """Handle the peer cluster relation changed event."""
+        if not (deployment_desc := self.charm.opensearch_peer_cm.deployment_desc()):
+            # Deployment description not available yet
+            # Nothing to do
+            return
+
+        if (
+            deployment_desc.typ != DeploymentType.MAIN_ORCHESTRATOR
+            and (rel_data := self.charm.opensearch_peer_cm.rel_data())
+            and rel_data.deployment_desc
+        ):
+            # Not the main orchestrator: set the profile from the peer-cluster
+            if self.apply(rel_data.deployment_desc.config.profile):
+                # We need to restart
+                self.charm.trigger_restart()
```

Comment:
why do we set the profile from the main orchestrator? every application should have its own profile.

Reply:
The profile also sets global, cluster-wide settings. I don't think we should have all the apps enforcing a given index template just because "data.cold" is being set to profile=production.

Reply:
That is something the main orch. must mandate instead.